File: android/app/src/test/java/com/localtrans/app/ui/transfers/TransferDisplayTest.kt
package com.localtrans.app.ui.transfers

import org.junit.Assert.*
import org.junit.Test

/**
 * 传输卡展示纯函数单测(M2 T5,移植桌面 transferDisplay.test.ts 口径)
 */
class TransferDisplayTest {

    private fun t(
        jobId: Long = 1,
        state: String = "active",
        total: Long = 1000,
        done: Long = 500,
        remoteDone: Long = 0,
        localRole: String = "receiver",
        direction: String = "rx",
        queuePos: Int? = null,
        startedAtMs: Long? = null,
        finishedAtMs: Long? = null
    ) = TransferUi(
        jobId = jobId, name = "f.bin", total = total, done = done, state = state,
        speedBps = 0, peer = "aa9988112233", direction = direction, localRole = localRole,
        progressPercent = 50, etaSecs = -1, failReason = "", localPath = null,
        remoteDone = remoteDone, instant = false,
        queuePos = queuePos, batchId = null, children = emptyList(),
        partsId = null, startedAtMs = startedAtMs, finishedAtMs = finishedAtMs,
        sourcePath = null
    )

    // ===== progressPair(R1 单进度条主进度口径) =====

    @Test
    fun `receiver main progress is done`() {
        val pair = progressPair(t(localRole = "receiver", done = 500))
        assertEquals(500L, pair.mainDone)
        assertEquals(0.0, pair.backlogPct, 0.001)
    }

    @Test
    fun `sender main progress is remote done`() {
        val pair = progressPair(t(localRole = "source-push", done = 900, remoteDone = 600))
        assertEquals(600L, pair.mainDone)
        assertEquals(30.0, pair.backlogPct, 0.001)
    }

    @Test
    fun `sender without mirror data falls back to done`() {
        // 旧任务无 remote_done 镜像:done>0 且 remote==0 且非终态 → 回退 done,绝不显示 0%
        val pair = progressPair(t(localRole = "source-push", done = 400, remoteDone = 0))
        assertEquals(400L, pair.mainDone)
    }

    @Test
    fun `sender terminal keeps remote as main`() {
        val pair = progressPair(
            t(localRole = "source-push", state = "done", done = 1000, remoteDone = 0)
        )
        assertEquals(0L, pair.mainDone)
    }

    // ===== senderDisplayState / isAwaitingConfirmText =====

    @Test
    fun `full but unconfirmed is awaiting`() {
        val state = senderDisplayState(
            t(localRole = "source-push", done = 1000, remoteDone = 800)
        )
        assertEquals(SenderDisplayState.AWAITING_CONFIRM, state)
    }

    @Test
    fun `backlog over 20 pct is backlog`() {
        val state = senderDisplayState(
            t(localRole = "source-push", done = 500, remoteDone = 100)
        )
        assertEquals(SenderDisplayState.BACKLOG, state)
    }

    @Test
    fun `awaiting text only for non terminal sender`() {
        val active = t(localRole = "source-push", done = 1000, remoteDone = 900)
        assertTrue(isAwaitingConfirmText(active))
        val receiver = t(localRole = "receiver", done = 1000, remoteDone = 900)
        assertFalse(isAwaitingConfirmText(receiver))
        val doneCard = t(
            localRole = "source-push", state = "done", done = 1000, remoteDone = 900
        )
        assertFalse(isAwaitingConfirmText(doneCard))
    }

    // ===== queueText =====

    @Test
    fun `queue text with position`() {
        assertEquals("排队中 · 第 3 位", queueText(3, "push"))
    }

    @Test
    fun `queue text fallback by direction`() {
        assertEquals("排队中，连接对端...", queueText(null, "pull"))
        assertEquals("等待对方接收...", queueText(null, "push"))
    }

    // ===== splitActiveHistory =====

    @Test
    fun `split separates active and history`() {
        val jobs = listOf(
            t(jobId = 1, state = "done"),
            t(jobId = 2, state = "active"),
            t(jobId = 3, state = "pending", queuePos = 2),
            t(jobId = 4, state = "pending", queuePos = 1),
            t(jobId = 5, state = "failed"),
            t(jobId = 6, state = "paused"),
            t(jobId = 7, state = "cancelling")
        )
        val split = splitActiveHistory(jobs)
        // active 排序:active > cancelling > pending(queue_pos 升序) > paused
        assertEquals(listOf(2L, 7L, 4L, 3L, 6L), split.active.map { it.jobId })
        // history 按 finished_at 降序(新的在前,null 靠后)
        assertEquals(setOf(1L, 5L), split.history.map { it.jobId }.toSet())
    }

    @Test
    fun `history sorted by finished desc nulls last`() {
        val jobs = listOf(
            t(jobId = 1, state = "done", finishedAtMs = 100),
            t(jobId = 2, state = "done", finishedAtMs = 300),
            t(jobId = 3, state = "done", finishedAtMs = null),
            t(jobId = 4, state = "done", finishedAtMs = 200)
        )
        val split = splitActiveHistory(jobs)
        assertEquals(listOf(2L, 4L, 1L, 3L), split.history.map { it.jobId })
    }

    // ===== elapsed / fingerprint / child state =====

    @Test
    fun `elapsed freezes at finish`() {
        val card = t(startedAtMs = 1000, finishedAtMs = 61000)
        assertEquals(60L, elapsedSeconds(card, nowMs = 999_000))
        assertNull(elapsedSeconds(t(startedAtMs = null), nowMs = 999_000))
    }

    @Test
    fun `fingerprint abbreviation`() {
        assertEquals("aa9988...112233", fingerprintAbbr("aa9988112233".repeat(2)))
        assertEquals("short", fingerprintAbbr("short"))
    }

    @Test
    fun `peer display name prefers name over fingerprint`() {
        val peers = listOf(PeerNameSource("AA99881122334455", "我的电脑", ""))
        assertEquals("我的电脑", peerDisplayName("aa99881122334455", peers))
        assertEquals("aa9988...334455", peerDisplayName("aa99881122334455", emptyList()))
    }

    @Test
    fun `child state text`() {
        assertEquals("完成", childStateText("done"))
        assertEquals("失败", childStateText("failed"))
        assertEquals("weird", childStateText("weird"))
    }
}

File: ui/src/lib/__tests__/transferDisplay.test.ts
import { describe, it, expect } from 'vitest'
import {
  peerDisplayName,
  progressPair,
  senderDisplayState,
  queueText,
  fingerprintAbbr,
  childRetryParams,
  splitActiveHistory,
} from '../transferDisplay'

describe('peerDisplayName', () => {
  const peers = [
    { fingerprint: 'AA9988CCDDEE1122334455667788990011223344', name: '我的手机', alias: '小米' },
    { fingerprint: 'FF11223344556677889900112233445566778899', name: '', alias: '' },
  ]

  it('别名优先', () => {
    expect(peerDisplayName('aa9988ccddee1122334455667788990011223344', peers)).toBe('小米')
  })
  it('无别名回退广播名', () => {
    const noAlias = [{ fingerprint: 'ff11223344556677889900112233445566778899', name: '华为平板', alias: '' }]
    expect(peerDisplayName('FF11223344556677889900112233445566778899', noAlias)).toBe('华为平板')
  })
  it('无信任记录回退指纹缩写', () => {
    expect(peerDisplayName('aabbccddeeff0011223344556677889900aabbcc', [])).toBe('aabbcc...aabbcc')
  })
  it('指纹缩写格式 aa9988..ccdd12 风格(前6...后6)', () => {
    const fp = 'aa99881122334455667788990011223344ccdd12'
    expect(peerDisplayName(fp, [])).toBe('aa9988...ccdd12')
  })
  it('短指纹原样返回', () => {
    expect(peerDisplayName('abc', [])).toBe('abc')
  })
  it('fingerprintAbbr 短指纹原样', () => {
    expect(fingerprintAbbr('abcd')).toBe('abcd')
  })
})

describe('progressPair', () => {
  it('source-push 主进度=remote_done, sentDone=done, 积压40%', () => {
    expect(
      progressPair({ local_role: 'source-push', done: 100, remote_done: 60, total: 100 })
    ).toEqual({ mainDone: 60, sentDone: 100, backlogPct: 40 })
  })
  it('接收方向 mainDone=done', () => {
    expect(
      progressPair({ local_role: 'destination', done: 70, remote_done: 999, total: 100 })
    ).toEqual({ mainDone: 70, sentDone: 70, backlogPct: 0 })
  })
  it('remote_done undefined 当 0(新发送任务)', () => {
    const r = progressPair({ local_role: 'source-push', done: 0, total: 100 })
    expect(r.mainDone).toBe(0)
    expect(r.backlogPct).toBe(0)
  })
  it('容错:done>0 且 remote_done=0 且 total>0 且非终态 → 视为无镜像数据退回单进度', () => {
    const r = progressPair({ local_role: 'source-push', done: 50, total: 100, state: 'active' })
    expect(r.mainDone).toBe(50)
    expect(r.backlogPct).toBe(0)
  })
  it('终态不触发无镜像容错(终态 remote_done=0 即真 0)', () => {
    const r = progressPair({ local_role: 'source-push', done: 50, total: 100, state: 'done' })
    expect(r.mainDone).toBe(0)
  })
})

describe('senderDisplayState', () => {
  it('满格未确认 → awaiting-confirm', () => {
    expect(senderDisplayState({ done: 100, remote_done: 80, total: 100, state: 'active' })).toBe('awaiting-confirm')
  })
  it('积压超 20% → backlog', () => {
    expect(senderDisplayState({ done: 40, remote_done: 10, total: 100, state: 'active' })).toBe('backlog')
  })
  it('积压恰 20% 不算 backlog', () => {
    expect(senderDisplayState({ done: 30, remote_done: 10, total: 100, state: 'active' })).toBe('normal')
  })
  it('正常 → normal', () => {
    expect(senderDisplayState({ done: 50, remote_done: 48, total: 100, state: 'active' })).toBe('normal')
  })
  it('remote_done undefined 非满格不算 backlog(旧任务容错)', () => {
    expect(senderDisplayState({ done: 30, total: 100, state: 'active' })).toBe('normal')
  })
})

describe('queueText', () => {
  it('queue_pos 非空 → 排队中 · 第 N 位', () => {
    expect(queueText(3, 'push')).toBe('排队中 · 第 3 位')
  })
  it('queue_pos null 回退 push 文案', () => {
    expect(queueText(null, 'push')).toBe('等待对方接收...')
  })
  it('queue_pos undefined 回退 pull 文案', () => {
    expect(queueText(undefined, 'pull')).toBe('排队中，连接对端...')
  })
})

describe('childRetryParams', () => {
  it('push-rel returns path+relDir from first valid item', () => {
    const req = { type: 'push-rel' as const, items: [['C:/dir/a.txt', 'sub'], ['C:/b.txt', '']] as [string, string][] }
    expect(childRetryParams(req)).toEqual({ path: 'C:/dir/a.txt', relDir: 'sub' })
  })

  it('push-rel with no valid item returns null', () => {
    expect(childRetryParams({ type: 'push-rel', items: [['C:/a.txt', '']] })).toBeNull()
  })

  it('push (absolute path) only needs path, relDir is empty string', () => {
    const req = { type: 'push' as const, paths: ['C:/abs/file.bin'] }
    expect(childRetryParams(req)).toEqual({ path: 'C:/abs/file.bin', relDir: '' })
  })

  it('push with empty paths returns null', () => {
    expect(childRetryParams({ type: 'push', paths: [] })).toBeNull()
  })

  it('pull uses req.path with empty relDir', () => {
    expect(childRetryParams({ type: 'pull', path: 'C:/d/f.mp4' })).toEqual({ path: 'C:/d/f.mp4', relDir: '' })
  })

  it('missing request returns null', () => {
    expect(childRetryParams(undefined)).toBeNull()
  })
})

describe('splitActiveHistory', () => {
  const mk = (over: Partial<import('../transferDisplay').TransferLike>): import('../transferDisplay').TransferLike => ({
    job_id: 'j',
    state: 'active',
    queue_pos: null,
    started_at_ms: 1000,
    finished_at_ms: null,
    ...over,
  })

  it('空输入返回两个空数组', () => {
    expect(splitActiveHistory([])).toEqual({ active: [], history: [] })
  })

  it('混合状态正确分组', () => {
    const jobs = [
      mk({ job_id: 'a', state: 'active' }),
      mk({ job_id: 'b', state: 'done' }),
      mk({ job_id: 'c', state: 'pending' }),
      mk({ job_id: 'd', state: 'paused' }),
      mk({ job_id: 'e', state: 'failed' }),
      mk({ job_id: 'f', state: 'interrupted' }),
      mk({ job_id: 'g', state: 'cancelling' }),
    ]
    const { active, history } = splitActiveHistory(jobs)
    expect(active.map((j) => j.job_id).sort()).toEqual(['a', 'c', 'd', 'g'])
    expect(history.map((j) => j.job_id).sort()).toEqual(['b', 'e', 'f'])
  })

  it('active 排序:active>cancelling>pending(queue_pos 升序,null 靠后)>paused', () => {
    const jobs = [
      mk({ job_id: 'paused1', state: 'paused' }),
      mk({ job_id: 'pend-null', state: 'pending', queue_pos: null }),
      mk({ job_id: 'canc', state: 'cancelling' }),
      mk({ job_id: 'pend-2', state: 'pending', queue_pos: 2 }),
      mk({ job_id: 'act', state: 'active' }),
      mk({ job_id: 'pend-1', state: 'pending', queue_pos: 1 }),
    ]
    const { active } = splitActiveHistory(jobs)
    expect(active.map((j) => j.job_id)).toEqual(['act', 'canc', 'pend-1', 'pend-2', 'pend-null', 'paused1'])
  })

  it('active 同级按 started_at_ms 升序(老的在前),null 靠后', () => {
    const jobs = [
      mk({ job_id: 'new', started_at_ms: 2000 }),
      mk({ job_id: 'null-ts', started_at_ms: null }),
      mk({ job_id: 'old', started_at_ms: 1000 }),
    ]
    const { active } = splitActiveHistory(jobs)
    expect(active.map((j) => j.job_id)).toEqual(['old', 'new', 'null-ts'])
  })

  it('history 按 finished_at_ms 降序(新的在前),null 靠后', () => {
    const jobs = [
      mk({ job_id: 'old', state: 'done', finished_at_ms: 1000 }),
      mk({ job_id: 'null-fin', state: 'failed', finished_at_ms: null }),
      mk({ job_id: 'new', state: 'done', finished_at_ms: 3000 }),
      mk({ job_id: 'mid', state: 'interrupted', finished_at_ms: 2000 }),
    ]
    const { history } = splitActiveHistory(jobs)
    expect(history.map((j) => j.job_id)).toEqual(['new', 'mid', 'old', 'null-fin'])
  })
})
